inwebo/inwebo-authenticator/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/InweboAuthenticator.java

```diff
@@ -0,0 +1,240 @@
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+
+package org.wso2.carbon.identity.authenticator;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.wso2.carbon.identity.application.authentication.framework.AbstractApplicationAuthenticator;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.authentication.framework.FederatedApplicationAuthenticator;
+import org.wso2.carbon.identity.application.authentication.framework.config.ConfigurationFacade;
+import org.wso2.carbon.identity.application.authentication.framework.context.AuthenticationContext;
+import org.wso2.carbon.identity.application.authentication.framework.exception.AuthenticationFailedException;
+import org.wso2.carbon.identity.application.common.model.ClaimMapping;
+import org.wso2.carbon.identity.application.common.model.Property;
+import org.apache.amber.oauth2.common.utils.JSONUtils;
+
+
+/**
+ * Authenticator of Inwebo
+ */
+public class InweboAuthenticator extends AbstractApplicationAuthenticator implements FederatedApplicationAuthenticator {
+
+    private static Log log = LogFactory.getLog(InweboAuthenticator.class);
+    private static final long serialVersionUID = -4154255583070524018L;
+    private String pushResponse = null;
+
+
+    /**
+     * Check whether the authentication or logout request can be handled by the authenticator
+     */
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+        if (log.isDebugEnabled()) {
+            log.debug("Inside InweboAuthenticator.canHandle()");
+        }
+        return (pushResponse != null && pushResponse.contains(InweboConstants.PUSHRESPONSE));
+    }
+
+    /**
+     * initiate the authentication request
+     */
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request,
+                                                 HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException {
+        try {
+            if (pushResponse == null) {
+                Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
+                if (authenticatorProperties != null) {
+                    String relyingParty = request.getParameterValues("relyingParty")[0];
+                    String type = request.getParameterValues("type")[0];
+                    String referer = request.getHeader("referer").toString();
+                    String finalReferer = "";
+                    if ((referer.charAt(referer.lastIndexOf(":") + 5) + "").equals("/")) {
```

Comment:
Why this logic is so hard coded with counting the index in exact numbers? It's not good to have it like this as even a slight change will break the flow.